### ISSUE TITLE
Leak in git_odb_backend_one_pack/git_odb_add_backend + new API request

### ISSUE THREAD

**Issue by lstoppa** [libgit2/libgit2#6844]
### Reproduction steps
This is quite hard to reproduce because I cannot, unless I modified the source code. So where I work we're playing with the packing API. We have a very simple code:

```
const char* path{ /* doesn't matter what*/ };
git_odb_backend* backend_pack{nullptr};
if (::git_odb_backend_one_pack(&backend_pack, path) < 0)
  return -1;
git_odb* pack_odb{nullptr};
if (::git_odb_new(&pack_odb) < 0) {
  ::git_odb_free(pack_obj);
  return -1;
}
if (::git_odb_add_backend(pack_odb, backend_pack, 0) < 0) {
  ::git_odb_free(pack_obj);
  return -1;
}
// now you can use pack_odb
```

### Expected behavior
When anything fals I would expect no memory leak.

### Actual behavior
We have multiple memory leaks depending on the failing path.
- if `::git_odb_backend_one_pack` fails, all is good, the implementation takes care of all the allocated memory and everything gets freed
- if `::git_odb_new` fails, unfortunately we leak `backend_pack` because there's no documented way of freeing this memory. After looking into the code, it seems like an undocumented way that works might be `backend_pack->free(backend_pack)`, but this is quite ugly. There are two problems in general with `::git_odb_backend_one_pack`:
1. First of all the public documentation doesn't say it allocates memory
2. Second, I think we should consider to add a new API:
```
void git_odb_backend_free(git_odb_backend* ptr) {
  GIT_ASSERT_ARG(ptr);
  ptr->free(ptr);
}
```
-  The biggest problem however is with `::git_odb_add_backend` where I believe we have multiple leaks. Please see [this](https://github.com/libgit2/libgit2/blob/main/src/libgit2/odb.c#L579).
This is a diff I have created:
```
diff --git a/src/libgit2/odb.c b/src/libgit2/odb.c
index fec1e45b9..ae1a03a7c 100644
--- a/src/libgit2/odb.c
+++ b/src/libgit2/odb.c
@@ -600,10 +600,13 @@ static int add_backend_internal(
 
        if (git_mutex_lock(&odb->lock) < 0) {
                git_error_set(GIT_ERROR_ODB, "failed to acquire the odb lock");
+               backend->free(backend);
+               git__free(internal);
                return -1;
        }
        if (git_vector_insert(&odb->backends, internal) < 0) {
                git_mutex_unlock(&odb->lock);
+               backend->free(backend);
                git__free(internal);
                return -1;
        }
```

As you can imagine, those cold paths are quite hard to hit (I am not sure when `git_vector_insert` really fails), nor I don't know how to cause the `git_mutex_lock` to fail, so maybe those leaks are non-existent. Still, I think an official way for destroying `git_odb_backend` objects is probably something we should consider.

### Version of libgit2 (release number or SHA1)
Develop

### Operating system(s) tested
Linux

**Comment by lstoppa** [comment]
No, we're not really hitting it :) The whole story is that we have c++ RAII wrapper over libgit2, and most of time if you really want to see if the code is correct is to inspect the code, and that's exactly what happened here. While checking the source code for libgit2 we realised that
1. there is a potential leak when `git_mutex_lock` and `git_vector_insert` happen, and
2. that we have no public API for freeing backends.

Now the question is: will point #1 really happen? Probably not, because I think `git_vector_insert` fails only when we have no more available memory, so probably we don't care. Regarding `git_mutex_lock` probably it's also another very specific situation, and maybe we don't care? 

So the leak here is only probably theoretical, no repro available. If you think this bug in practice doesn't happen, please feel free to close it!

**Comment by ethomson** [comment]
Thanks for the details; I'll leave this open because I do want to fix this.

1. The libgit2 philosophy is (or perhaps _was_) - if an allocation error occurs, we are well and truly hosed and shouldn't bother trying to clean up because it wouldn't do any good. _Well, unless it's a **really big** allocation, then we probably should._ And now we've taken a rule that should be objective and made it subjective. It's harder for developers to reason about, _and_ the tools that we use to catch memory leaks give us false positives. Our lives would be simpler if we just said "we clean up on any failure, even allocations". So we should.
2. We should definitely have a way to free what you create. 🙏

**Comment by lstoppa** [comment]
Like always, if you feel like we can help with this bug (like we did with the FIPS code), let us know: we can just create probably a couple of MRs that will address each issue independently
- the theoretical leak (a straightforward change I think)
- the additional public API interface required
